### ISSUE TITLE
Meson has_link_argument doesn't work for --version-script on FreeBSD

### ISSUE THREAD

**Issue by lantw44** [mesonbuild/meson#3923]
Meson `has_link_argument` links an executable to test linker arguments. However, it doesn't work for `--version-script` on FreeBSD with the default linker because GNU BFD ld can only use `--version-script` when linking a shared library. GNU BFD ld on other operating systems (Linux, NetBSD, OpenBSD), or other linkers (GNU gold, LLVM lld) on FreeBSD, don't seem to have the problem. FreeBSD has a plan to switch to LLVM lld, but currently the default linker is still GNU BFD ld.

```
Running compile:
Working directory:  /tmp/tmpyij_lcod
Command line:  clang -B/home/lantw44/.local/bin -L/home/lantw44/gnome/devinstall/lib -L/usr/local/lib /tmp/tmpyij_lcod/testfile.c -pipe -D_FILE_OFFSET_BITS=64 -o /tmp/tmpyij_lcod/output.exe -O0 -Wl,--fatal-warnings -Wl,--version-script,/home/lantw44/gnome/source/gvfs/client/symbol.map 

Code:
 int main(int argc, char **argv) { return 0; }
Compiler stdout:

Compiler stderr:
 /home/lantw44/.local/bin/ld: /tmp/tmpyij_lcod/output.exe: local symbol `environ' in /usr/lib/crt1.o is referenced by DSO
/home/lantw44/.local/bin/ld: final link failed: Bad value
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Compiler for C supports link arguments -Wl,--version-script,/home/lantw44/gnome/source/gvfs/client/symbol.map: NO
```

This problem was found when testing gvfs:

https://bugzilla.gnome.org/show_bug.cgi?id=794365#c19
https://gitlab.gnome.org/GNOME/gvfs/merge_requests/5#note_269420

The only workaround I found was to manually add `-shared` to the linker command line when testing `--version-script`, but I guess it is going to cause problems if there is a compiler which doesn't understand `-shared`.

**Comment by nirbheek** [comment]
This can't be fixed by changing `cc.has_link_argument()`. This needs to be fixed by adding a new kwarg to build targets for the version script because only Meson can know the appropriate conditions to use it.

**Comment by lantw44** [comment]
I agree that this feature should not be put into `cc.has_link_argument`, but it will be nice if both `cc.has_argument` and `cc.has_link_argument` can print warnings for `-Wl,--version-script` after the feature is implemented.

**Comment by nirbheek** [comment]
> but it will be nice if both cc.has_argument and cc.has_link_argument can print warnings for -Wl,--version-script after the feature is implemented.

Completely agree.

**Comment by yurikoles** [comment]
bump?
